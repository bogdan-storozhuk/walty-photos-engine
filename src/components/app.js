import React from 'react';
import { Route, Switch } from 'react-router-dom';
import 'bootstrap/dist/css/bootstrap.min.css';

import { MainPage, PhotoPageContainer } from '.';

const App = () => (
  <Switch>
    <Route path="/" component={MainPage} exact />
    <Route
      path="/photo-page/:id"
      render={({ match }) => {
        const { id } = match.params;
        return <PhotoPageContainer photoId={id} />;
      }}
      exact
    />
    <Route
      path="/:tag1?/:tag2?/:tag3?/:tag4?/:tag5?/:tag6?/:tag7?/:tag8?/:tag9?/:tag10?"
      render={({ match }) => <MainPage tags={match.params} />}
    />
  </Switch>
);

App.propTypes = {};

export default App;
