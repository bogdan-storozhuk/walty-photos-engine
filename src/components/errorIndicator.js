import React from 'react';

const ErrorIndicator = () => <div>Error!</div>;

ErrorIndicator.propTypes = {};

export default ErrorIndicator;
