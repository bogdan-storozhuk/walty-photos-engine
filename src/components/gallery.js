import React from 'react';
import { useSelector } from 'react-redux';
import PropTypes from 'prop-types';

import { GalleryItem } from './';
import { Spinner } from './';
import { ErrorIndicator } from './';

export const Gallery = ({ photos }) => (
  <div className="row mt-3">
    {photos.map((photo) => (
      <GalleryItem key={photo.id} photo={photo} />
    ))}
  </div>
);

const GalleryContainer = () => {
  const photos = useSelector((state) => state.photoReducer.photos);
  const loading = useSelector((state) => state.photoReducer.loading);
  const error = useSelector((state) => state.photoReducer.error);

  if (loading) {
    return <Spinner />;
  }

  if (error) {
    return <ErrorIndicator />;
  }

  return <Gallery photos={photos} />;
};

Gallery.propTypes = {
  photos: PropTypes.arrayOf(
    PropTypes.shape({
      largeImageURL: PropTypes.string,
      id: PropTypes.number,
    })
  ),
};

export default GalleryContainer;
