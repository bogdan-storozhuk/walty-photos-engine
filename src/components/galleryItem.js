import React from 'react';
import PropTypes from 'prop-types';

import { GalleryItemContainer, GalleryItemImage } from '../styled/';

const GalleryItem = ({ photo }) => (
  <GalleryItemContainer className="col-lg-3 col-md-4 col-6 mb-4">
    <GalleryItemImage
      src={photo.largeImageURL}
      alt={photo.id}
    ></GalleryItemImage>
  </GalleryItemContainer>
);

GalleryItem.propTypes = {
  photo: PropTypes.shape({
    largeImageURL: PropTypes.string,
    id: PropTypes.number,
  }),
};

export default GalleryItem;
