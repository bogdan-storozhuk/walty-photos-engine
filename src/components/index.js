import App from './app';
import ErrorBoundry from './errorBoundry';
import ErrorIndicator from './errorIndicator';
import GalleryContainer, { Gallery } from './gallery';
import GalleryItem from './galleryItem';
import LastSearchesPanel from './lastSearchesPanel';
import MainPage from './mainPage';
import SearchPanel from './searchPanel';
import Spinner from './spinner';
import Tag from './tag';
import PhotoPage from './photoPage';

export {
  App,
  MainPage,
  SearchPanel,
  LastSearchesPanel,
  GalleryContainer,
  Gallery,
  GalleryItem,
  Tag,
  ErrorBoundry,
  ErrorIndicator,
  Spinner,
  PhotoPage,
};
