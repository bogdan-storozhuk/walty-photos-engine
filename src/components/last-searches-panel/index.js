import LastSearchesPanel from './last-searches-panel';
export default LastSearchesPanel;