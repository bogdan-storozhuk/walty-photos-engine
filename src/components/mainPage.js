import React, { useEffect } from 'react';
import { useDispatch } from 'react-redux';
import Container from 'react-bootstrap/Container';
import PropTypes from 'prop-types';

import {
  loadTagsFromURL,
  photoLoadStart,
  lastSearchedTagAdded,
} from '../ducks';
import { SearchPanel } from '.';
import { LastSearchesPanel } from '.';
import { Gallery } from '.';
import {
  deleteUndefinedValuesFromObject,
  transformObjectIntoArray,
} from '../utils/utils';

const MainPage = ({ tags }) => {
  const dispatch = useDispatch();
  useEffect(() => {
    if (tags) {
      const lastSearchedTags = { ...localStorage };
      deleteUndefinedValuesFromObject(tags);
      const lastSearchedTagsArray = transformObjectIntoArray(lastSearchedTags);
      const tagArray = transformObjectIntoArray(tags);
      dispatch(lastSearchedTagAdded(lastSearchedTagsArray));
      dispatch(loadTagsFromURL(tagArray));
      dispatch(photoLoadStart(tagArray));
    }
  }, [dispatch, tags]);
  return (
    <Container>
      <SearchPanel />
      <LastSearchesPanel />
      <Gallery />
    </Container>
  );
};

MainPage.propTypes = {
  tags: PropTypes.shape({
    tag1: PropTypes.string,
    tag2: PropTypes.string,
    tag3: PropTypes.string,
    tag4: PropTypes.string,
    tag5: PropTypes.string,
    tag6: PropTypes.string,
    tag7: PropTypes.string,
    tag8: PropTypes.string,
    tag9: PropTypes.string,
    tag10: PropTypes.string,
  }),
};

export default MainPage;
