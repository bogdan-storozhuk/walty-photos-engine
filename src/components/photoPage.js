import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import Container from 'react-bootstrap/Container';
import PropTypes from 'prop-types';

import { singlePhotoStart } from '../ducks/';
import { Spinner, ErrorIndicator } from '../components/';
import { StyledImageContainer, StyledImage } from '../styled/';

export const PhotoPage = ({ singlePhoto }) => (
  <Container className="mt-4">
    <StyledImageContainer>
      <StyledImage
        className="img-fluid mx-auto d-block"
        src={singlePhoto.largeImageURL}
        alt={singlePhoto.id}
      />
    </StyledImageContainer>
    <div className="table-responsive mt-2">
      <table className="table table-bordered">
        <thead>
          <tr>
            <th>Downloads</th>
            <th>Favorites</th>
            <th>Likes</th>
            <th>Views</th>
            <th>Tags</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>{singlePhoto.downloads}</td>
            <td>{singlePhoto.favorites}</td>
            <td>{singlePhoto.likes}</td>
            <td>{singlePhoto.views}</td>
            <td>{singlePhoto.tags}</td>
          </tr>
        </tbody>
      </table>
    </div>
  </Container>
);

PhotoPage.propTypes = {
  singlePhoto: PropTypes.shape({
    id: PropTypes.number,
    downloads: PropTypes.number,
    favorites: PropTypes.number,
    likes: PropTypes.number,
    views: PropTypes.number,
    tags: PropTypes.string,
    largeImageURL: PropTypes.string,
  }),
};

const PhotoPageContainer = ({ photoId }) => {
  const dispatch = useDispatch();
  const singlePhoto = useSelector((state) => state.photoReducer.searchedPhoto);
  const loading = useSelector((state) => state.photoReducer.loading);
  const error = useSelector((state) => state.photoReducer.error);

  useEffect(() => {
    if (photoId) {
      dispatch(singlePhotoStart(photoId));
    }
  }, [dispatch, photoId]);

  if (loading) {
    return <Spinner />;
  }

  if (error) {
    return <ErrorIndicator />;
  }

  if (!singlePhoto) {
    return null;
  }

  return <PhotoPage singlePhoto={singlePhoto} />;
};

PhotoPageContainer.propTypes = {
  photoId: PropTypes.string,
};

export default PhotoPageContainer;
