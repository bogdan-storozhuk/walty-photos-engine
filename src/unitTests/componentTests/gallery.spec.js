import React from 'react';
import * as ReactReduxHooks from 'react-redux';
import configureMockStore from 'redux-mock-store';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, configure } from 'enzyme';
import { GalleryContainer, Gallery } from '../../components';

configure({ adapter: new Adapter() });

let useEffect;

const mockUseEffect = () => {
  useEffect.mockImplementationOnce((f) => f());
};

const setUpMockStore = (state) => {
  let store = configureMockStore([])(state);
  /* mocking useEffect */
  useEffect = jest.spyOn(React, 'useEffect');
  mockUseEffect(); // 2 times
  mockUseEffect(); //
  /* mocking useSelector on our mock store */
  jest
    .spyOn(ReactReduxHooks, 'useSelector')
    .mockImplementation((state) => store.getState());
  /* mocking useDispatch on our mock store  */
  jest
    .spyOn(ReactReduxHooks, 'useDispatch')
    .mockImplementation(() => store.dispatch);
  /* shallow rendering */
  return store;
};

describe('Gallery Component', () => {
  const props = {
    photos: [],
  };
  describe('Gallery Component initial', () => {
    const gallery = shallow(<Gallery {...props} />);

    it('render initial', () => {
      expect(gallery.find('GalleryItem')).toHaveLength(0);
    });
  });
});

describe('GalleryContainer Component', () => {
  const initialState = {
    photoReducer: {
      searchedPhoto: null,
      photo: [],
      loading: false,
      error: null,
    },
  };
  describe('GalleryContainer Loading', () => {
    const nextState = {
      ...initialState,
      photoReducer: {
        ...initialState.photoReducer,
        loading: true,
      },
    };
    const store = setUpMockStore(nextState);
    const wrapper = shallow(<GalleryContainer store={store} />);

    it('Renders Spinner', () => {
      expect(wrapper.find('Spinner')).toHaveLength(1);
    });

    it('not render <ErrorIndicator />', () => {
      expect(wrapper.find('ErrorIndicator')).toHaveLength(0);
    });

    it('not render <Gallery/>', () => {
      expect(wrapper.find('Gallery')).toHaveLength(0);
    });

    it('renders properly', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('GalleryContainer with errorMsg', () => {
    const nextState = {
      ...initialState,
      photoReducer: {
        ...initialState.photoReducer,
        error: 'some error',
      },
    };
    console.log(nextState);
    const store = setUpMockStore(nextState);
    const wrapper = shallow(<GalleryContainer store={store} />);

    it('Renders ErrorIndicator', () => {
      expect(wrapper.find('ErrorIndicator')).toHaveLength(1);
    });
    it('Renders Spinner', () => {
      expect(wrapper.find('Spinner')).toHaveLength(0);
    });
  });
});
