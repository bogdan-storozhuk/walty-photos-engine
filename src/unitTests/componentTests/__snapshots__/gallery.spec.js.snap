// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GalleryContainer Component GalleryContainer Loading renders properly 1`] = `ShallowWrapper {}`;
